core/src/connection/handshake.rs

```diff
@@ -82,22 +82,29 @@ impl<T: AsyncRead + AsyncWrite> Future for Handshake<T> {
 }
 
 fn client_hello<T: AsyncWrite>(connection: T, gc: Vec<u8>) -> WriteAll<T, Vec<u8>> {
-    let packet = protobuf_init!(ClientHello::new(), {
-        build_info => {
-            product: protocol::keyexchange::Product::PRODUCT_PARTNER,
-            platform: protocol::keyexchange::Platform::PLATFORM_LINUX_X86,
-            version: 0x10800000000,
-        },
-        cryptosuites_supported => [
-            protocol::keyexchange::Cryptosuite::CRYPTO_SUITE_SHANNON,
-        ],
-        login_crypto_hello.diffie_hellman => {
-            gc: gc,
-            server_keys_known: 1,
-        },
-        client_nonce: util::rand_vec(&mut thread_rng(), 0x10),
-        padding: vec![0x1e],
-    });
+    let packet = {
+        let mut msg = ClientHello::new();
+        {
+            let mut msg = msg.mut_build_info();
+            msg.set_product(protocol::keyexchange::Product::PRODUCT_PARTNER);
+            msg.set_platform(protocol::keyexchange::Platform::PLATFORM_LINUX_X86);
+            msg.set_version(0x10800000000);
+            msg
+        };
+        {
+            let repeated = msg.mut_cryptosuites_supported();
+            repeated.push(protocol::keyexchange::Cryptosuite::CRYPTO_SUITE_SHANNON)
+        };
+        {
+            let mut msg = msg.mut_login_crypto_hello().mut_diffie_hellman();
+            msg.set_gc(gc);
+            msg.set_server_keys_known(1);
+            msg
+        };
+        msg.set_client_nonce(util::rand_vec(&mut thread_rng(), 0x10));
+        msg.set_padding(vec![0x1e]);
+        msg
```

Comment:
Can you rewrite those in a more "top level" style, as I used in https://github.com/librespot-org/librespot/issues/129#issuecomment-363982143
It will be more concise, even if a bit repetitive